### PR TITLE
Jsx to tsx with functionalities

### PR DESCRIPTION
**Pull request by PKolur**
Jsx to tsx with functionalities

### CODE REVIEW

File: components/ColorPicker.tsx [hunk 1]
```diff
@@ -3,14 +3,13 @@ Make this application dynamic to change color on slider movement.
...
 
 export default function () {
-    const { r, g, b } = { r: 30, g: 30, b: 30 };
-
-    const getVal = (e, val) => {
-        console.log(e.target.name, val);
-    };
+    const [red, setRed] = useState(30);
+    const [green, setGreen] = useState(30);
+    const [blue, setBlue] = useState(30);
```
Comment (anjan-kp) [inline components/ColorPicker.tsx:12]: It can be handled using one useState instead of 3

File: components/ColorPicker.tsx [hunk 2]
```diff
@@ -30,7 +29,7 @@ export default function () {
                         orientation="vertical"
                         defaultValue={30}
                         valueLabelDisplay="on"
-                        onChange={getVal}
+                        onChange={(e, val)=>setRed(Number(val))}
```
Comment (anjan-kp) [inline components/ColorPicker.tsx:32]: one handler function can be used to get value instead of 3 different functions.

File: components/Layout.tsx [hunk 3]
```diff
@@ -6,16 +6,11 @@ import ColorPicker from "./ColorPicker";
...
-    const handleChange = (event, newValue) => {
-        setValue(newValue);
-    };
-
     return (
         <Box sx={{ width: "100%", typography: "body1" }}>
             <TabContext value={value}>
                 <Box sx={{ borderBottom: 1, borderColor: "divider" }}>
-                    <TabList onChange={handleChange} aria-label="lab API tabs example">
+                    <TabList onChange={((e, newValue)=>setValue(newValue))} aria-label="lab API tabs example">
```
Comment (anjan-kp) [inline components/Layout.tsx:13]: Is this change required?

File: components/PhoneBook.tsx [hunk 4]
```diff
@@ -1,36 +1,59 @@
...
+    }
+    const addRecord = (e: React.FormEvent<HTMLFormElement>) => {
         e.preventDefault();
+        console.log('event :', formData);
+        phoneContext?.savePhoneList(formData);
+        setFormData({name:'', phone:''});
     };
 
     return (
-        <form autocomplete="off" onSubmit={addRecord} style={style.form.container}>
```
Comment (anjan-kp) [inline components/PhoneBook.tsx:19]: autocomplete="off" to be changed to autoComplete="off"

File: components/PhoneBook.tsx [hunk 5]
```diff
@@ -1,36 +1,59 @@
...
-            <input style={style.form.inputs} name="n" type="text" />
+            <input style={style.form.inputs} name="name" type="text" value={formData.name} onChange={handleOnChange} />
             <br />
             <label>Phone:</label>
             <br />
-            <input style={style.form.inputs} name="p" type="text" />
+            <input style={style.form.inputs} name="phone" type="text" value={formData.phone} onChange={handleOnChange} />
             <br />
-            <input style={style.form.submitBtn} type="submit" value="Add User" />
+            <input style={style.form.submitBtn} type="submit" value="Add User" disabled={formData && formData.name && formData.phone ? false : true} />
```
Comment (anjan-kp) [inline components/PhoneBook.tsx:42]: Use object destructuring. No need to mention formData.name, formData.phone etc.

File: components/PhoneBook.tsx [hunk 6]
```diff
@@ -1,36 +1,59 @@
...
+    const phoneContext = useContext(PhoneContext);
+    let sortedArray = phoneContext?.phoneDetails?.sort((a,b)=>{
+        if (a.name < b.name)
+           return -1;
+        else if (a.name == b.name)
+           return 0;
+        else
+           return 1;
+    });
     return (
```
Comment (anjan-kp) [inline components/PhoneBook.tsx:57]: Create a separate function for sorting which will take parameter/s and sort based on that.

File: pages/_app.tsx [hunk 7]
```diff
@@ -1,8 +1,11 @@
+import { PhoneContextProvider } from '../components/context/PhoneBookContext'
 import '../styles/globals.css'
 import type { AppProps } from 'next/app'
 
 function MyApp({ Component, pageProps }: AppProps) {
-  return <Component {...pageProps} />
+  return <PhoneContextProvider>
+    <Component {...pageProps} />
+  </PhoneContextProvider>
```
Comment (anjan-kp) [inline pages/_app.tsx:8]: Context is not required on the app level. It's only needed for the phonebook component.

File: components/context/PhoneBookContext.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    useEffect(() => {
+        const data = localStorage.getItem(LOCAL_KEY);
+        if (data) {
+            setPhoneDetails(JSON.parse(data))
+        }
+    }, [])
+
+    useEffect(() => {
+        localStorage.setItem('data', JSON.stringify(phoneDetails))
```
Comment (anjan-kp) [inline components/context/PhoneBookContext.tsx:36]: create a custom hook useStorage to read/write to storage.

File: components/PhoneBook.tsx [hunk 9]
```diff
@@ -1,36 +1,59 @@
...
             <br />
-            <input style={style.form.submitBtn} type="submit" value="Add User" />
+            <input style={style.form.submitBtn} type="submit" value="Add User" disabled={formData && formData.name && formData.phone ? false : true} />
         </form>
     );
 };
 
 const InformationTable = () => {
+    const phoneContext = useContext(PhoneContext);
+    let sortedArray = phoneContext?.phoneDetails?.sort((a,b)=>{
```
Comment (anjan-kp) [inline components/PhoneBook.tsx:49]: why are you using phonecontext.phoneDetails ? Use destructuring always.

File: components/ColorPicker.tsx (hunk 1 above)
Comment (PKolur) [reply components/ColorPicker.tsx]: Done

File: components/context/PhoneBookContext.tsx (hunk 8 above)
Comment (PKolur) [reply components/context/PhoneBookContext.tsx:36]: I will add the created custom hook

File: components/ColorPicker.tsx [hunk 10]
```diff
@@ -3,14 +3,20 @@ Make this application dynamic to change color on slider movement.
...
-
-    const getVal = (e:any, val:any) => {
-        console.log(e.target.name, val);
-    };
+    const [colorPallets, setcolorPallets] = useState({
+        red:30,
+        green:30,
+        blue:30
+    });
+    const getVal = (e: Event, val: number | number[]) =>{
```
Comment (vineeth-ktp) [inline components/ColorPicker.tsx:15]: you can mention return type, if nothing is returning, you can use :void like below:

const functionName = (props: Any):void =>{}

File: components/PhoneBook.tsx [hunk 11]
```diff
@@ -1,70 +1,160 @@
...
+Session storage custome hook for persiting the data.
 */
+import { useState, useContext, createContext, useEffect } from "react";
 import Wrapper from "./Wrapper";
 import { style } from "../styles/style";
+import {useLocalStorage} from '../hooks/useStorage';
+
+
+// type Contexting and creating.
+type phoneListProps = {
```
Comment (vineeth-ktp) [inline components/PhoneBook.tsx:18]: variable name should follow same standards - **phoneListProps** can be **PhoneListProps**

File: components/PhoneBook.tsx [hunk 12]
```diff
@@ -1,70 +1,160 @@
...
+        phone:''
+    });
+    const phonesList = useContext(PhonesBookContext);
+    const [errors, setErrors] = useState({nameError:'', phoneError:''});
+    const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        setFormData({...formData, [e.target.name]:e.target.value});
+    }
+    const addRecord = (e:  React.SyntheticEvent) => {
         e.preventDefault();
+        console.log('Error :', errors);
```
Comment (vineeth-ktp) [inline components/PhoneBook.tsx:42]: remove console log statements

File: components/PhoneBook.tsx [hunk 13]
```diff
@@ -1,70 +1,160 @@
...
+            phonesList?.savePhonesList(formData);
+            setFormData({name:'', phone:''});
+            setErrors({nameError:'', phoneError:''});
+        }
     };
-
+    const validate = () => {
+        const {name, phone} = formData;
+        let error = {nameError:'', phoneError: ''};
+        if (!phonesList) return null;
```
Comment (vineeth-ktp) [inline components/PhoneBook.tsx:52]: Line 52 and 53 can be replaced with ternary operator.

File: components/PhoneBook.tsx [hunk 14]
```diff
@@ -1,70 +1,160 @@
...
+            const { phoneList } = phonesList;
+            console.log('phoneList :', phoneList);
```
Comment (vineeth-ktp) [inline components/PhoneBook.tsx:54]: remove console log statement

File: components/PhoneBook.tsx [hunk 15]
```diff
@@ -1,70 +1,160 @@
...
+    if (a.val < b.val)
+       return -1;
+    else if (a.val == b.val)
+       return 0;
+    else
+       return 1;
+});
 const InformationTable = () => {
+    const phonesContextList = useContext(PhonesBookContext);
+    console.log('phonesContextList:', phonesContextList);
```
Comment (vineeth-ktp) [inline components/PhoneBook.tsx:94]: remove console log statements

File: components/PhoneBook.tsx [hunk 16]
```diff
@@ -1,70 +1,160 @@
...
-    <>
-        <FormWrapper />
-        <br />
-        <InfoWrapper />
-    </>
-);
+export default () => {
+    const [phoneDetails, setPhoneDetails] = useState<phoneListProps[]>([]);
+    const [getLocalItems, setLocalItems] = useLocalStorage();
+    console.log('storage : ', useLocalStorage(), getLocalItems)
```
Comment (vineeth-ktp) [inline components/PhoneBook.tsx:138]: remove console log statement

File: components/PhoneBook.tsx [hunk 17]
```diff
@@ -1,70 +1,160 @@
...
+            const { phoneList } = phonesList;
+            console.log('phoneList :', phoneList);
+        let isValid = true;
+        let pattern = new RegExp(/^[0-9\b]+$/);
+        let namePattern = new RegExp(/^[a-zA-Z]+$/);
+        if(name && (!namePattern.test(name)) || (phoneList?.length && phoneList?.some(val=>val.name === name.toUpperCase()))){            
+            isValid =  false;
+            error.nameError = `Invalid!!!, Name ${namePattern.test(name) ? "exists already":"should be in alphabetic"}.`
+        }
+        if(phone && (!pattern.test(phone) || phone.length !== 10)){
```
Comment (vineeth-ktp) [inline components/PhoneBook.tsx:62]: instead of checking phone errors, use HTML attribute in `title="Error Message" pattern="[1-9]{1}[0-9]{9}"` if it helps

**Review by vineeth-ktp — COMMENTED** [review]
check the review comments

File: components/PhoneBook.tsx (hunk 17 above)
Comment (vineeth-ktp) [reply components/PhoneBook.tsx:62]: If for checking already exist in phonebook similar to name - then just use pattern to limit only 10 digit numbers

Comment (PKolur) [reply components/PhoneBook.tsx:60]: checking before submitting and triggering error if input on phone number is not valid

Comment (vineeth-ktp) [reply components/PhoneBook.tsx:60]: Not good enough, you can avoid the whole compile time check and logic if you can restrict at input level. That is HTML basics right, you can use it.

File: components/PhoneBook.tsx [hunk 18]
```diff
@@ -49,9 +48,8 @@ const PhoneBookForm = () => {
...
         const {name, phone} = formData;
         let error = {nameError:'', phoneError: ''};
-        if (!phonesList) return null;
-            const { phoneList } = phonesList;
-            console.log('phoneList :', phoneList);
+        // if (!phonesList) return null;
+        const { phoneList } = phonesList!;
         let isValid = true;
         let pattern = new RegExp(/^[0-9\b]+$/);
         let namePattern = new RegExp(/^[a-zA-Z]+$/);
```
[the remark this answers is not in the document]
Comment (PKolur) [reply components/PhoneBook.tsx:55]: ok
